File: data_structures/typescript/src/LinkedList/SinglyLinkedList.ts
interface INode<T> {
  data: T;
  next: INode<T> | null;
}

export class SinglyLinkedList<T> {
  #_Node = class Node<T> implements INode<T> {
    public next: Node<T> | null = null;
    constructor(public data: T) {}
  };

  #_head: INode<T> | null = null;
  #_length = 0;

  constructor(init?: T | T[], ...rest: T[]) {
    if (Array.isArray(init)) {
      init.forEach(this.append, this);
    } else if (init) {
      [init, ...rest].forEach(this.append, this);
    }
  }

  get length() {
    return this.#_length;
  }

  /**
   * adds given element to end of list
   * @param data
   */
  append(data: T) {
    const node = new this.#_Node(data);
    if (this.#_head === null) {
      this.#_head = node;
    } else {
      let current = this.#_head;
      while (current.next !== null) {
        current = current.next;
      }
      current.next = node;
    }
    this.#_length++;
  }

  /**
   * removes first found instance of given element
   * @param data
   * @returns true if data found and removed, else false
   */
  remove(data: T) {
    if (this.#_head === null) {
      return false;
    }
    if (this.#_head.data === data) {
      this.#_head = this.#_head.next;
      this.#_length--;
      return true;
    }
    let current = this.#_head;
    while (current.next !== null) {
      if (current.next.data === data) {
        current.next = current.next.next;
        this.#_length--;
        return true;
      }
      current = current.next;
    }
    return false;
  }

  clear() {
    this.#_head = null;
    this.#_length = 0;
  }

  public toString() {
    let result = '';
    let current = this.#_head;
    while (current != null) {
      result += current.data;
      if (current.next !== null) {
        result += ' -> ';
      }
      current = current.next;
    }
    return result;
  }

  public toArray() {
    const result = [] as T[];
    let current = this.#_head;
    while (current !== null) {
      result.push(current.data as T);
      current = current.next;
    }
    return result;
  }
}

File: data_structures/linked_list/singly_linked_list.py
class ListNode:
    def __init__(self, data):
        self.data = data
        self.next: ListNode = None


class SinglyLinkedList:
    def __init__(self) -> None:
        self._head: ListNode = None
        self._length = 0

    def __str__(self) -> str:
        return " -> ".join(map(str, self.to_list()))

    @property
    def length(self) -> int:
        """
        Returns
            length of the list
        """
        return self._length

    def to_list(self) -> list:
        """
        Returns
            The elements of the LinkedList in a python list
        """
        r = []
        n = self._head
        while n is not None:
            r.append(n.data)
            n = n.next
        return r

    def clear(self) -> None:
        """
        Clears all elements from the list
        """
        self._head = None
        self._length = 0

    def append(self, data) -> None:
        """
        Add the given value to the end of the list

        Params
            data: the value to append
        """
        node = ListNode(data)
        if self._head is None:
            self._head = node
        else:
            curr = self._head
            while curr.next is not None:
                curr = curr.next
            curr.next = node
        self._length += 1

    def remove(self, data) -> bool:
        """
        Removes first found instance of given value

        Params
            data: The value to remove

        Returns
            True if element removed, else False
        """
        if self._head is None:
            return False

        if self._head.data == data:
            self._head = self._head.next
            self._length -= 1
            return True

        curr = self._head
        while curr.next is not None:
            if curr.next.data == data:
                curr.next = curr.next.next
                self._length -= 1
                return True
            curr = curr.next
        return False
